src/Elcodi/Component/Plugin/Services/PluginManager.php

```diff
@@ -57,252 +69,194 @@ class PluginManager
+                    $pluginNamespace,
+                    $pluginConfiguration
                 );
 
-                $plugins->set($bundleNamespace, $specification);
-            }
-        }
+            $this->savePlugin($pluginInstance);
 
-        /**
-         * We remove old plugin references
-         */
-        foreach ($plugins as $pluginNamespace => $plugin) {
-            if (!in_array($pluginNamespace, $bundlesFound)) {
+            $pluginsLoaded[] = $pluginInstance;
+            if (isset($plugins
+                [$pluginNamespace])) {
                 unset($plugins[$pluginNamespace]);
             }
         }
 
-        $pluginsArray = $plugins->toArray();
-
-        $this
-            ->configurationManager
-            ->set('store.plugins', $pluginsArray);
+        /**
+         * Every Plugin instance inside $plugins array should be removed from
+         * database, because they are not longer installed
+         */
+        $this->removePlugins($plugins);
 
-        return $pluginsArray;
+        return $pluginsLoaded;
     }
 
     /**
-     * Read plugin specification
+     * Load existing plugins from database and return an array with them all,
+     * indexed by its namespace
      *
-     * @param string $bundlePath Bundle path
-     *
-     * @return array Plugin specification
+     * @return Plugin[] Plugins indexed by namespace
      */
-    protected function getPluginSpecification($bundlePath)
+    protected function getExistingPlugins()
     {
-        $yaml = new Parser();
-        $specificationFilePath = $bundlePath . '/plugin.yml';
-        if (!file_exists($specificationFilePath)) {
-            return [];
+        $pluginsIndexed = [];
+        $plugins = $this
+            ->pluginRepository
+            ->findAll();
+
+        /**
+         * @var Plugin $plugin
+         */
+        foreach ($plugins as $plugin) {
+            $pluginNamespace = $plugin->getNamespace();
+            $pluginsIndexed[$pluginNamespace] = $plugin;
         }
 
-        return array_intersect_key(
-            $yaml->parse(file_get_contents($specificationFilePath)),
-            array_flip([
-                'name',
-                'author',
-                'url',
-                'description',
-                'year',
-                'version',
-                'fa_icon',
-                'configuration_route',
-                'enabled',
-                'visible',
-            ])
-        );
+        return $pluginsIndexed;
     }
 
     /**
-     * Get plugins
+     * Load installed plugin bundles and return an array with them, indexed by
+     * their namespaces
      *
-     * @return array Plugins
+     * @return PluginInterface[]|Bundle[] Plugins installed
      */
-    public function getPlugins()
+    protected function getInstalledPluginBundles()
     {
-        if ($this->plugins) {
-            return $this->plugins;
-        }
+        $plugins = [];
+        $bundles = $this
```

Comment:
If there's no other need for the `Kernel` to be injected, maybe better pass directly `kernel.bundles` as a parameter.